### PR TITLE
Long-form bootstrapping syntax allowing multiple apps, root components

### PR DESCRIPTION
**Pull request by alxhub**

### CODE REVIEW

**Comment by alxhub** [comment]
This is ready for an initial review while I verify that it will actually build and run - I expect there will be some changes to get the tests passing.

**Comment by alxhub** [comment]
Having lots of issues with the tests but debugging is yielding good results. I'm a little worried about backwards compatibility with bootstrap(), I think it will suffer, but it *is* an architectural change.

File: modules/angular2/src/core/application_common.ts [hunk 1]
```diff
@@ -77,63 +77,69 @@ import {internalView} from 'angular2/src/core/compiler/view_ref';
...
-        .toValue(DOM.defaultDoc()),
-    bind(APP_COMPONENT).toValue(appComponentType),
+    bind(APP_COMPONENT).toValue(componentType),
     bind(APP_COMPONENT_REF_PROMISE)
         .toFactory(
-            (dynamicComponentLoader, injector, testability, registry) => {
+            (dynamicComponentLoader, injector) => {
               // TODO(rado): investigate whether to support bindings on root component.
-              return dynamicComponentLoader.loadAsRoot(appComponentType, null, injector)
+              return dynamicComponentLoader.loadAsRoot(componentType, null, injector)
```
Comment (tbosch) [inline modules/angular2/src/core/application_common.ts]: I think this should be split up: Dont' call `DynamicComponentLoader.loadAsRoot`, but implement the logic that is in `DynamicComponentLoader.loadAsRoot` directly here:

1. call `Compiler.compileHost`, which is asynchronous
2. call `ViewManager.createRootHostView` with the result, which is synchronous

Could we have the new bootstrap adjust for this?

E.g. have an `platform.compile(componentType):Promise<ProtoViewRef>` and a `platform.application(componentType, protoViewRef, bindings):ComponentRef` and a
`platform.compileAndCreate(componentType, bindings):Promise<ComponentRef>`.

Comment (tbosch) [inline modules/angular2/src/core/application_common.ts]: This separation is important for serverside rendering, as there we compile once, cache the result, and create the root component multiple times (on every request).

**Comment by alxhub** [comment]
I've solved the problem I was having with `bootstrap()` backwards compatibility: `bootstrap()` accepts the user's bindings for the application, but can't determine which bindings belong in the application injector and which ones belong in the component. Some bindings, like `Router`, need to exist in the component injector because they rely on the current component, while others such as an application API service should be shared among all of an app's root components.

A solution to this problem is to introduce `Platform.simpleApplication(type, bindings)`, which returns a `Promise<RootComponentRef>` and uses a single injector but only supports one root component. `bootstrap()` then accepts any bindings and doesn't run into the same scoping problem.

**Comment by mhevery** [comment]
@alxhub Awesome. Do you have some work in progress which you can share, so that we can start the review process?

**Comment by alxhub** [comment]
@mhevery This PR is the full bootstrapping change that's ready for review. I'm still having some issues with the Dart tests but I can work through those separately.

**Comment by alxhub** [comment]
@mhevery: Looks like I can't add the label, but it should be good to merge. Feel free to add it if you think the PR is ready to go in.

**Pull request closed without merging**